### PR TITLE
Timing/master

### PR DESCRIPTION
**Pull request by karasevb**

### CODE REVIEW

File: ompi/util/timings.h [hunk 1]
```diff
@@ -4,123 +4,215 @@
...
+    int inum;
+    int cnt;
+    int error;
+    int enabled;
+    opal_timing_ts_func_t get_ts;
+    ompi_timing_list_t *timing;
+} ompi_timing_env_t;
+
+#define OMPI_TIMIG_ENV_ITEM_IDX(idx) ({                                        \
+    __label__ exit;                                                            \
```
Comment (artpol84) [inline ompi/util/timings.h:32]: Is this a standard feature?
or GCC extension?

File: ompi/util/timings.h [hunk 2]
```diff
@@ -4,123 +4,215 @@
...
+            tmp->item = NULL;                                                  \
+        }                                                                      \
+        free(tmp);                                                             \
+        tmp = NULL;                                                            \
+    } while ( NULL != t);                                                      \
+exit:;                                                                         \
+    OMPI_TIMING_env.cnt = 0;                                                   \
+})
+
+#define OMPI_TIMING_INIT(_inum)                                                \
```
Comment (artpol84) [inline ompi/util/timings.h:96]: Also - OMPI timings are not env-based.
So no need in "env" in the name.

Comment (artpol84) [inline ompi/util/timings.h]: Look how it is done for opal

File: ompi/util/timings.h [hunk 3]
```diff
@@ -4,123 +4,215 @@
...
+    OMPI_TIMING_env.ts = OMPI_TIMING_env.get_ts();                             \
+    OMPI_TIMING_env.timing = (ompi_timing_list_t*)malloc(sizeof(ompi_timing_list_t));       \
+    OMPI_TIMING_env.timing->next = NULL;                                       \
+    OMPI_TIMING_ITEM_INIT(&OMPI_TIMING_env.timing->item);                      \
+
+#define OMPI_TIMING_NEXT(fmt, ...) {                                           \
+    ompi_timing_item_t *item;                                                  \
+    if( OMPI_TIMING_env.cnt &&                                                 \
+                !(OMPI_TIMING_env.cnt % OMPI_TIMING_env.inum) ){               \
+        OMPI_TIMING_ENV_ITEM_EXTEND;                                           \
```
Comment (artpol84) [inline ompi/util/timings.h:112]: I'd say that extend can be inline function - it is rarely called

File: ompi/util/timings.h [hunk 4]
```diff
@@ -4,123 +4,215 @@
...
+    OMPI_TIMING_env.error = 0;                                                 \
+    OMPI_TIMING_env.ts = OMPI_TIMING_env.get_ts();                             \
+    OMPI_TIMING_env.timing = (ompi_timing_list_t*)malloc(sizeof(ompi_timing_list_t));       \
+    OMPI_TIMING_env.timing->next = NULL;                                       \
+    OMPI_TIMING_ITEM_INIT(&OMPI_TIMING_env.timing->item);                      \
+
+#define OMPI_TIMING_NEXT(fmt, ...) {                                           \
+    ompi_timing_item_t *item;                                                  \
+    if( OMPI_TIMING_env.cnt &&                                                 \
+                !(OMPI_TIMING_env.cnt % OMPI_TIMING_env.inum) ){               \
```
Comment (artpol84) [inline ompi/util/timings.h:111]: For safety - lets check for >:
`OMPI_TIMING_env.cnt > cur_size` => we need to reallocate.

File: ompi/util/timings.h [hunk 5]
```diff
@@ -4,123 +4,215 @@
...
+        OMPI_TIMING_ENV_ITEM_EXTEND;                                           \
+    }                                                                          \
+    item = OMPI_TIMIG_ENV_ITEM_IDX(OMPI_TIMING_env.cnt);                       \
+    if ( NULL == item ) {                                                      \
+        printf("OMPI_TIMING [%d %s]: item not found\n",                        \
```
Comment (artpol84) [inline ompi/util/timings.h:116]: Not clear what does it mean, I guess this means that allocation failed. if so - we need to check it when extending

File: ompi/util/timings.h (hunk 1 above)
Comment (karasevb) [reply ompi/util/timings.h:32]: This is GCC extension, it allows to declare local labels. More info: https://gcc.gnu.org/onlinedocs/gcc/Local-Labels.html

File: ompi/util/timings.h [hunk 6]
```diff
@@ -4,123 +4,215 @@
...
+    int cnt = OPAL_TIMING_ENV_CNT(func);                                       \
+    int i;                                                                     \
+    for(i = 0; i < cnt; i++){                                                  \
+        char *desc;                                                            \
+        double ts = OPAL_TIMING_ENV_GETDESC_PREFIX(_prefix, func, i, &desc);   \
+        OMPI_TIMING_APPEND(desc, ts);                                          \
+    }                                                                          \
 }
 
+#define OMPI_TIMING_IMPORT_OPAL(func) \
```
Comment (artpol84) [inline ompi/util/timings.h:157]: This is easier to keep as inline function.
It is out of the critical path.

File: ompi/util/timings.h [hunk 7]
```diff
@@ -4,123 +4,215 @@
...
-    }                                                         \
+#define OMPI_TIMING_APPEND(desc,ts) {                                          \
+    ompi_timing_item_t *item;                                                  \
+    if( OMPI_TIMING_env.cnt && !(OMPI_TIMING_env.cnt % OMPI_TIMING_env.inum)){ \
+        OMPI_TIMING_ENV_ITEM_EXTEND;                                           \
+    }                                                                          \
+    item = OMPI_TIMIG_ENV_ITEM_IDX(OMPI_TIMING_env.cnt);                       \
+    if ( NULL == item ) {                                                      \
+        printf("OMPI_TIMING [%d %s]: item not found\n",                        \
+            __LINE__, __FUNCTION__);                                           \
```
Comment (artpol84) [inline ompi/util/timings.h:137]: Lets continue to print filename as it was in original implememtation.

Comment (artpol84) [inline ompi/util/timings.h]: !!!!

File: ompi/util/timings.h [hunk 8]
```diff
@@ -4,123 +4,215 @@
...
+                        OMPI_TIMING_env.get_ts() - OMPI_TIMING_env.ts;         \
+    asprintf(&item->desc[OMPI_TIMING_env.cnt % OMPI_TIMING_env.inum],          \
+            fmt, ## __VA_ARGS__);                                              \
+    item->use++;                                                               \
+    OMPI_TIMING_env.cnt++;                                                     \
+    OMPI_TIMING_env.ts = OMPI_TIMING_env.get_ts();                             \
 }
 
-#define OMPI_TIMING_APPEND(desc,ts) {                                   \
-    char *ptr = strrchr(__FILE__, '/');                           \
```
Comment (artpol84) [inline ompi/util/timings.h:56]: !!!!!

File: ompi/runtime/ompi_mpi_init.c [hunk 9]
```diff
@@ -976,6 +977,7 @@ int ompi_mpi_init(int argc, char **argv, int requested, int *provided)
      * and clear timing structure */
     OMPI_TIMING_NEXT("barrier-finish");
     OMPI_TIMING_OUT;
+    OMPI_TIMING_ENV_FINALIZE;
```
Comment (artpol84) [inline ompi/runtime/ompi_mpi_init.c]: lets merge OMPI_TIMING_ENV_FINALIZE and OMPI_TIMING_OUT

File: ompi/util/timings.h [hunk 10]
```diff
@@ -4,123 +4,215 @@
...
+    ompi_timing_item_t *item;
+    struct ompi_timing_list_t *next;
+} ompi_timing_list_t;
+
+typedef struct {
+    double ts;
+    const char *prefix;
+    int inum;
+    int cnt;
+    int error;
```
Comment (artpol84) [inline ompi/util/timings.h:25]: int size, to keep allocated space availble

File: ompi/util/timings.h [hunk 11]
```diff
@@ -4,123 +4,215 @@
...
+    int enabled;
+    opal_timing_ts_func_t get_ts;
+    ompi_timing_list_t *timing;
```
Comment (artpol84) [inline ompi/util/timings.h:28]: ` ompi_timing_list_t *timing_cur;` - keep the pointer to the last item so you wont need to look for it every time

File: ompi/util/timings.h [hunk 12]
```diff
@@ -4,123 +4,215 @@
...
+    i = (idx % OMPI_TIMING_env.inum);                                          \
+    item = t->item;                                                            \
+exit:                                                                          \
+    item;                                                                      \
+})
+
+#define OMPI_TIMING_ITEM_INIT(item) {                                          \
+    *item = (ompi_timing_item_t*)malloc(sizeof(ompi_timing_item_t));           \
+    (*item)->in = (double*)malloc(sizeof(double) * OMPI_TIMING_env.inum);      \
+    (*item)->desc = (char**)malloc(sizeof(char**) * OMPI_TIMING_env.inum);     \
```
Comment (artpol84) [inline ompi/util/timings.h:51]: doesn't seem right - `sizeof(char**)`, should be `sizeof(char*)`

File: ompi/util/timings.h [hunk 13]
```diff
@@ -4,123 +4,215 @@
...
+                !(OMPI_TIMING_env.cnt % OMPI_TIMING_env.inum) ){               \
+        OMPI_TIMING_ENV_ITEM_EXTEND;                                           \
+    }                                                                          \
+    item = OMPI_TIMIG_ENV_ITEM_IDX(OMPI_TIMING_env.cnt);                       \
+    if ( NULL == item ) {                                                      \
+        printf("OMPI_TIMING [%d %s]: item not found\n",                        \
+            __LINE__, __FUNCTION__);                                           \
+        abort();                                                               \
+    }                                                                          \
+    item->in[OMPI_TIMING_env.cnt % OMPI_TIMING_env.inum] =                     \
```
Comment (artpol84) [inline ompi/util/timings.h:120]: I don't want extra memory allocation.
I want description field to be of fixed size, but we check for overflow. as it was directed in the draft.

File: ompi/util/timings.h (hunk 1 above)
Comment (artpol84) [reply ompi/util/timings.h:32]: So we want the code to be portable, no GCC extensions.

**Pull request closed without merging**